renku/service/views/error_handlers.py

```diff
@@ -0,0 +1,403 @@
+            raise
+
+    return decorated_function
+
+
+@handle_config_read_errors
+def handle_config_write_errors(f):
+    """Wrapper which handles setting config errors."""
+    # noqa
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        """Represents decorated function."""
+        try:
+            return f(*args, **kwargs)
+        except ParameterError as e:
+            if "Invalid parameter value" in str(e):
+                parameter = None
+                match = re.search('Key "(.*)" not found', str(e))
+                if len(match.groups()) > 0:
+                    parameter = match.group(1)
+                raise IntermittentSettingExistsError(e, parameter)
+            raise
+
+    return decorated_function
+
+
+def handle_datasets_write_errors(f):
+    """Wrapper which handles datasets write errors."""
+    # noqa
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        """Represents decorated function."""
+        try:
+            return f(*args, **kwargs)
+        except DatasetImageError as e:
+            error_message = str(e)
+            if "Duplicate dataset image" in error_message:
+                raise UserDatasetsMultipleImagesError(e)
+            elif "couldn't be mirrored" in error_message:
+                raise UserDatasetsUnreachableImageError(e)
+            raise
+        except ValidationError as e:
+            items = squash(e.messages).items()
+            for key, value in items:
+                if "".join(value) == "Field may not be null.":
+                    raise UserMissingFieldError(e, key)
+            raise
+        except DatasetExistsError as e:
+            raise IntermittentDatasetExistsError(e)
+        except RenkuException as e:
+            if str(e).startswith("invalid file reference"):
+                raise IntermittentFileNotExistsError(e)
+            raise
+
+    return decorated_function
+
+
+def handle_datasets_unlink_errors(f):
+    """Wrapper which handles datasets unlink errors."""
+    # noqa
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        """Represents decorated function."""
+        try:
+            return f(*args, **kwargs)
+        except RenkuException as e:
+            if str(e).startswith("Invalid parameter") and "No records found" in str(e):
+                raise UserDatasetsUnlinkError(e)
+            raise
+        except ValueError as e:
+            if "one of the filters must be specified" in str(e):
+                raise UserDatasetsUnlinkError(e)
+            raise
+
+    return decorated_function
+
+
+def handle_migration_errors(f):
+    """Wrapper which handles migrations exceptions."""
+    # noqa
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        """Represents decorated function."""
+        try:
+            return f(*args, **kwargs)
+        except (TemplateUpdateError, DockerfileUpdateError, MigrationError) as e:
+            raise UserOutdatedProjectError(e)
```

Comment:
This means a migration failed. Not that a migration is required. So telling they user they should migrate when they just tried this and it failed might not be the best thing to do :smile:

Reply:
Ah right! I reused an error originally meant to supersede `MigrationRequired`, not `MigrationError` :facepalm: